subsys/mgmt/ec_host_cmd: rework Host Command support

niedzwiecki-dawid opened the pull request:
Rework the Host Command support. It includes:
-update of naming
-change API to backend
-chagne a way of definding rx and tx buffers
-fix synchronization between the handler and backend layer -simply the HC handler
-update the directory structure

Signed-off-by: Dawid Niedzwiecki <dn@semihalf.com>

niedzwiecki-dawid committed:
subsys/mgmt/ec_host_cmd: rework Host Command support

Rework the Host Command support. It includes:
-change API to backend
-change a way of defining rx and tx buffers
-fix synchronization between the handler and backend layer
-simplify the HC handler

Signed-off-by: Dawid Niedzwiecki <dawidn@google.com>

semihalf-barnas-michal commented on subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_espi.c:
```diff
@@ -0,0 +1,135 @@
+	struct espi_callback espi_cb;
+};
+
+#define EC_HOST_CMD_ESPI_DEFINE(_name)                                                             \
```
I'm not sure if it's a good idea to dereference non-existing yet object, even if it's inside macro that will be used later. If you remove the API requirement to know specific drivers, it won't be needed.

semihalf-barnas-michal replied:
Also I don't see the need for this macro, if it's only used once with one name

semihalf-niedzwiecki-dawid replied:
I think it is not a problem if it is within one file.

semihalf-niedzwiecki-dawid replied:
I think using macro is a good approach when you create a few variables that depend on their names.

semihalf-barnas-michal commented on subsys/mgmt/ec_host_cmd/ec_host_cmd_handler.c:
```diff
@@ -4,26 +4,41 @@
+				.buf = COND_CODE_1(CONFIG_EC_HOST_CMD_HANDLER_TX_BUFFER_DEF,       \
+						   (_name##_tx_buffer), (NULL)),                   \
+				.len_max =                                                         \
+					COND_CODE_1(CONFIG_EC_HOST_CMD_HANDLER_TX_BUFFER_DEF,      \
```
If you allow to use multiple backends, how do you define the buffer size? Multiple backends can have different requirements for the size, should the user just know that this buffer must be big enough for every backend?

semihalf-niedzwiecki-dawid replied:
Every backend defines default buffers sizes. If multiple backends are use, user has to set the config to meet requirements of the largest one. I updated the help message.

semihalf-barnas-michal commented on subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_espi.c:
```diff
@@ -0,0 +1,135 @@
+	uint16_t event_type = (uint16_t)espi_evt.evt_details;
+	/* tx stores the shared memory buf pointer and size, so use it */
+	const struct ec_host_cmd_request_header *rx_header = hc_espi->tx->buf;
+	const size_t shared_size = hc_espi->tx->len_max;
```
Regarding the comment of size for different backends. The eSPI can have a different shared memory size, if it's smaller than the "maximum" buffer size, it may overflow.

semihalf-niedzwiecki-dawid replied:
I don't think so, the sizes are checked before memcpy.

semihalf-barnas-michal commented on subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_espi.c:
```diff
@@ -0,0 +1,135 @@
+	 */
+	if (rx_header->prtcl_ver != 3 ||
+	    rx_valid_data_size > CONFIG_EC_HOST_CMD_HANDLER_RX_BUFFER ||
+	    rx_valid_data_size > shared_size) {
```
Aren't these 2 values equal?

semihalf-niedzwiecki-dawid replied:
No, `rx_valid_data_size` is amount of bytes to copy based on a header, `CONFIG_EC_HOST_CMD_HANDLER_RX_BUFFER` is the rx buffer size where data are copied to, `shared_size` is a size of the eSPI memory, where data are copied from.

semihalf-barnas-michal commented on subsys/mgmt/ec_host_cmd/backends/Kconfig:
```diff
@@ -3,61 +3,40 @@
 	bool "Host commands support using eSPI bus"
 	depends on ESPI_PERIPHERAL_EC_HOST_CMD
 	depends on ESPI_PERIPHERAL_CUSTOM_OPCODE
-	depends on DT_HAS_ZEPHYR_EC_HOST_CMD_PERIPH_ESPI_ENABLED
```
It should depend on the espi0 node being enabled

semihalf-niedzwiecki-dawid replied:
I don't think relaying on nodelabel is a good idea.

cybekRT replied:
espi0 was just an example, I meant that it should depend on any compatible node being enabled. Current standard for Kconfigs is to use default y and depends on node being enabled. After removing this line, you can enable host commands on espi, without having any espi node enabled, if I understand correctly.

semihalf-barnas-michal commented on subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_espi.c:
```diff
@@ -0,0 +1,135 @@
+	const struct device *const dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_host_cmd_backend));
+	struct ec_host_cmd_espi_config config = {.dev = dev};
+
+	ec_host_cmd_init(ec_host_cmd_backend_get_espi(), (void *)&config);
```
If I understand correctly, if you do not set a specific backend as chosen, then you have to manually call the init function with manually selected backend. I think that this custom code should have a pointers to available devices, probe them or do other magic and then call the ec_host_cmd_init using pointer to the device. Generic API shouldn't have any knowledge about the drivers implementation.

semihalf-niedzwiecki-dawid replied:
First of all, it is not a driver :-) a user(application code) doesn't use the host command API. The user only has to set the `DT_CHOSEN(zephyr_host_cmd_backend)` node, or call the `ec_host_cmd_init` function manually. `ec_host_cmd_backend_get_*`  functions are created for that purpose. The shell subsystem has pretty similar mechanism.
Anyway, I added comments about `DT_CHOSEN(zephyr_host_cmd_backend)`  to header files.

semihalf-barnas-michal commented on subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c:
```diff
@@ -66,17 +64,21 @@ struct shi_it8xxx2_cfg {
 	uint8_t out_msg[SPI_TX_MAX_FIFO_SIZE] __aligned(4);
 };
 
+struct ec_host_cmd_shi_ite {
```
It should have some suffix to inform if this is config, context or something else

semihalf-niedzwiecki-dawid replied:
Good point. Added.

semihalf-barnas-michal commented on subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c:
```diff
@@ -129,6 +138,7 @@ static void shi_ite_response_host_data(const struct device *dev, uint8_t *out_ms
 {
 	struct shi_it8xxx2_data *data = dev->data;
 
+	/* TODO sprawdzic czy to moze byc wolane w przerwaniu */
```
:)

semihalf-niedzwiecki-dawid replied:
Done.

semihalf-barnas-michal commented on subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c:
```diff
@@ -184,18 +194,17 @@ static int shi_ite_periph_send(const struct device *dev,
+	 */
 	for (int i = 0; i < EC_SHI_PAST_END_LENGTH; i++) {
-		data->out_msg[sizeof(out_preamble) + buf->len + i] = EC_SHI_PAST_END;
+		data->out_msg[sizeof(out_preamble) + data->tx->len + i] = EC_SHI_PAST_END;
```
Would it make sense to create a pointer as local variable instead of doing double dereference every loop?

semihalf-niedzwiecki-dawid replied:
I don't think so. It would be duplicating available data.

semihalf-niedzwiecki-dawid commented on include/zephyr/mgmt/ec_host_cmd/backend.h:
```diff
@@ -0,0 +1,149 @@
+/**
+ * @brief Config for the eSPI backend
+ */
+struct ec_host_cmd_espi_config {
```
I crated a common part for all configs that should be at the top of the custom config structure.

semihalf-niedzwiecki-dawid commented on include/zephyr/mgmt/ec_host_cmd/backend.h:
```diff
@@ -0,0 +1,149 @@
+	ec_host_cmd_backend_api_send send;
+};
+
+struct ec_host_cmd_backend *ec_host_cmd_backend_get_espi(void);
```
I don't get why it removes the 'generic' aspect of the backends. This function only returns a pointer to espi backend to pass it to the `ec_host_cmd_init` function.

cybekRT replied:
Because you have to implement it for every backend available, and adding new backend will require to add similar function for this backend. If you would have a function "ec_host_cmd_backend_get", that takes some pointer to device, as all APIs are doing, then you would have a generic function, independent of any backend.

semihalf-niedzwiecki-dawid replied:
If the `zephyr_host_cmd_backend` is not chosen, how would you know which backend is connected to this device. You need to call `ec_host_cmd_init` first to pass the device pointer as config to a backed. The backend pointer is required to call the `ec_host_cmd_init`, so it is a loop. You need a way to get pointer to a backend. The backends don't have their nodes dts, so there is no separated devices structures, with their API. The devices have peripheral API, not the backend API. The backend API is not available for application code. Anyway, this approach is common in Zephyr e.g. for [shell](https://github.com/zephyrproject-rtos/zephyr/blob/22ff66f2401e57e887322dacbe0d5d5f5f4322aa/include/zephyr/shell/shell_uart.h).

semihalf-niedzwiecki-dawid commented on include/zephyr/mgmt/ec_host_cmd/simulator.h:
```diff
@@ -25,8 +25,11 @@
  * will override the first callback installation.
  *
  * @param cb Callback that is called when device would send data to host.
+ * @param tx_buf Pointer of a poniter to the tx buf structure where data will
```
Done.

semihalf-niedzwiecki-dawid commented on include/zephyr/mgmt/ec_host_cmd/ec_host_cmd.h:
```diff
@@ -212,6 +219,8 @@ enum ec_host_cmd_status {
 	EC_HOST_CMD_MAX = UINT16_MAX /* Force enum to be 16 bits. */
 } __packed;
 
+int ec_host_cmd_init(struct ec_host_cmd_backend *backend, const void *transport_config);
```
Added.

semihalf-niedzwiecki-dawid commented on subsys/mgmt/ec_host_cmd/Kconfig:
```diff
@@ -15,24 +15,63 @@ config EC_HOST_CMD
 config EC_HOST_CMD_HANDLER_STACK_SIZE
 	int "Stack size for the EC host command handler thread"
-	default 800
+	default 650
```
The value is based on my tests. But I think it it better to keep it `800` and adjust per board

semihalf-niedzwiecki-dawid commented on subsys/mgmt/ec_host_cmd/Kconfig:
```diff
@@ -15,24 +15,63 @@ config EC_HOST_CMD
+	default 256
+	help
+	  Buffer size in bytes for TX buffer defined by the handler. Some backend layers
+	  can define their own buffer, so the size can be zero no to duplilate buffers.
```
Done.

semihalf-niedzwiecki-dawid commented on subsys/mgmt/ec_host_cmd/Kconfig:
```diff
@@ -15,24 +15,63 @@ config EC_HOST_CMD
+	default 60
+	range 0 99
+	help
+	  Initialization priority for Host Command. This must be bigger than the initialization
```
Done.

niedzwiecki-dawid committed:
subsys/mgmt/ec_host_cmd: update directory structure

The Host Commands can be used with different transport layers e.g. SHI
or eSPI. The code that provides the peripheral API and allows sending
and receiving Host Commands via different transport layers is not
actually drivers of a peripheral, so move it to the
subsys/mgmt/ec_host_cmd folder.

Signed-off-by: Dawid Niedzwiecki <dawidn@google.com>

niedzwiecki-dawid committed:
subsys/mgmt/ec_host_cmd: rename peripheral to backend

Follow naming pattern in the subsystems(logging or shell) and name
the layer between generic handler and peripheral driver "backend".

The name doesn't suit that well to the SHI backend, because there isn't
SHI API itself and the SHI interface is used only for the host
communication. So the backend code includes the peripheral driver itself.

Signed-off-by: Dawid Niedzwiecki <dawidn@google.com>

nordicjm commented on boards/posix/native_posix/native_posix.dts:
```diff
@@ -20,7 +20,6 @@
 		zephyr,flash = &flash0;
 		zephyr,entropy = &rng;
 		zephyr,flash-controller = &flashcontroller0;
-		zephyr,ec-host-interface = &hcp;
```
Typos in commit message: chagne = change, definding = defining, simply = simplify

semihalf-niedzwiecki-dawid replied:
Thanks, done.

nordicjm commented on doc/services/device_mgmt/ec_host_cmd.rst:
```diff
@@ -0,0 +1,51 @@
+The host command communication requires buffers for rx and tx. The buffers are be provided by the
+general handler if ``EC_HOST_CMD_HANDLER_RX_BUFFER/EC_HOST_CMD_HANDLER_TX_BUFFER>0``. The shared
+buffers are useful for applications that use multiple backends. Defining own buffers by every
+backend would increase the memory usage. However, some buffers can be defined by a peripheral driver e.g. eSPI. These ones should be reused as much as possible.
```
split into 2 lines

semihalf-niedzwiecki-dawid replied:
Done.

nordicjm commented on subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_espi.c:
```diff
@@ -4,103 +4,133 @@
-
-		espi_write_lpc_request(ESPI_DEVICE, ECUSTOM_HOST_CMD_SEND_RESULT, &res);
-		return;
+	/* Only support version 3 and make sure the number of bytes to copy is not
```
As someone that has no knowledge on this, what is version 3, why is only version 3 supported and should this be listed elsewhere e.g. Kconfig, documentation, etc.?

semihalf-niedzwiecki-dawid replied:
I added comments to Kconfig and documentation.

semihalf-niedzwiecki-dawid commented on CODEOWNERS:
```diff
@@ -259,7 +259,6 @@
 /drivers/dma/*pl330*                      @raveenp
 /drivers/dma/*iproc_pax*                  @raveenp
 /drivers/dma/*intel_adsp*                 @teburd @abonislawski
-/drivers/ec_host_cmd_periph/              @jettr
```
Done, thanks.

semihalf-niedzwiecki-dawid commented on boards/posix/native_posix/native_posix.dts:
```diff
@@ -156,9 +156,9 @@
 		compatible = "zephyr,native-posix-counter";
 	};
 
-	hcp: ec-host-cmd-periph {
```
Done, thanks.

semihalf-niedzwiecki-dawid commented on doc/develop/api/overview.rst:
```diff
@@ -81,7 +81,7 @@ between major releases are available in the :ref:`zephyr_release_notes`.
      - 1.14
 
-   * - :ref:`ec_host_cmd_periph_api`
+   * - :ref:`ec_host_cmd_backend_api`
```
That's true. Unfortunately, we haven't been able to test the API yet.

nordicjm commented:
Tried configuring hello_world for the `it8xxx2` board then checked with `menuconfig` at these options, and it seems `EC_HOST_CMD_BACKEND_SHI_ITE` is not available for this board despite the name suggesting that it is - is this expected?

semihalf-niedzwiecki-dawid commented:
> Tried configuring hello_world for the `it8xxx2` board then checked with `menuconfig` at these options, and it seems `EC_HOST_CMD_BACKEND_SHI_ITE` is not available for this board despite the name suggesting that it is - is this expected?

Have you enabled the `"ite,it8xxx2-shi"` node?

nordicjm reviewed (commented):
That was it, thanks.

nordicjm commented on subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c:
```diff
@@ -184,18 +193,17 @@ static int shi_ite_backend_send(const struct device *dev,
-	memcpy(data->out_msg + sizeof(out_preamble), buf->buf, buf->len);
-
-	/* Append our past-end byte, which we reserved space for. */
+	/* Data to send are already at out_msg + sizeof(out_preamble).
```
```suggestion
	/* Data to send size is: out_msg + sizeof(out_preamble).
```

nordicjm reviewed (approved):
+1 from me after these changes. I don't know anything about this system so in terms of reviewing the way the code works, I don't know on that front.

nordicjm reviewed (changes requested):
*that should have been request changes

semihalf-niedzwiecki-dawid commented:
> +1 from me after these changes. I don't know anything about this system so in terms of reviewing the way the code works, I don't know on that front.

Thanks a lot for the review.

fabiobaltieri commented on include/zephyr/mgmt/ec_host_cmd/backend.h:
```diff
@@ -21,98 +21,111 @@
 #endif
 
+struct ec_host_cmd_backend {
+	/** API provided by the backed.
+	 */
```
nit: this and the one below should fit in a line

fabiobaltieri commented on include/zephyr/mgmt/ec_host_cmd/backend.h:
```diff
@@ -21,98 +21,111 @@
-	 *  command framework and handler (when handler_owns). Buffer is owned
-	 *  by devices and lives as long as device is valid. Device will never
-	 *  read from this buffer (for security reasons).
+	/** Buffer to hold received data. The buffer is provided by the handler if
```
nit: Can't quite find a codingstyle entry but other code seems to be aligning multi line comments by leaving the first line with `/**` and nothing else, should this code stick to that convention as well?

fabiobaltieri commented on doc/services/device_mgmt/ec_host_cmd.rst:
```diff
@@ -0,0 +1,56 @@
+Architecture
+************
+The Host Command subsystem contains a few components:
+ * backend
+ * general handler
+ * command handler
```
nit: uppercase on the fist letter

fabiobaltieri commented on doc/services/device_mgmt/ec_host_cmd.rst:
```diff
@@ -0,0 +1,56 @@
+
+The host command subsystem should be initialized automatically during boot process, if a user choose
+periheral node with ``zephyr,host-cmd-backend`` in the dts files. In this case, it is up to a chosen
+backend to call the ``ec_host_cmd_init`` function.
```
`:c:func:ec_host_cmd_init` (also at line 39)

fabiobaltieri commented on include/zephyr/mgmt/ec_host_cmd/backend.h:
```diff
@@ -21,98 +21,111 @@
+ *		  are used to send data by the backend with the
+ *		  ec_host_cmd_backend_api_send function.
  *
  * @retval 0 if successful
```
These look a bit funky in the generated documentation, I think because you changed the indentation. You should align the "0 if successful" with the other entries so that the doc renders as "Return values <break> 0 if successful" rather than "Retval 0: <break> if successful" (see https://builds.zephyrproject.io/zephyr/pr/54132/docs/services/device_mgmt/ec_host_cmd.html#c.ec_host_cmd_backend_api_init)

fabiobaltieri replied:
Oh nevermind then. The link: click on "Documentation Build / Documentation Build (HTML)"  details -> gear icon -> "view job summary"

fabiobaltieri commented on subsys/mgmt/ec_host_cmd/ec_host_cmd_handler.c:
```diff
@@ -6,24 +6,39 @@
+	static K_KERNEL_STACK_DEFINE(_name##stack, CONFIG_EC_HOST_CMD_HANDLER_STACK_SIZE);         \
+	static struct k_thread _name##thread;                                                      \
+	static struct ec_host_cmd _name = {                                                        \
+		.rx_ctx =                                                                          \
+			{                                                                          \
```
keep the brace inline? same below

keith-zephyr replied:
I believe Fabio's comment still applies.

semihalf-niedzwiecki-dawid replied:
I know, but I think a custom format is not required here.

fabiobaltieri replied:
Well, your name is the maintainer one for this after this PR so your call I guess, but I think this is just plain wrong compared the the coding style. I'd say shut clang off and indent manually.

fabiobaltieri reviewed (commented):
Added few nitpicks for the doc format, having a real hard time following the changes in the backends in the third patch, those 4 points in the commit should really be at least 4 separate patches. Are you planning on maintaining this after this PR? If so, I'd say let's go with it but maybe add a MAINTAINERS entry for this.

semihalf-niedzwiecki-dawid commented:
> Added few nitpicks for the doc format, having a real hard time following the changes in the backends in the third patch, those 4 points in the commit should really be at least 4 separate patches. Are you planning on maintaining this after this PR? If so, I'd say let's go with it but maybe add a MAINTAINERS entry for this.

I get your point. The problem is that some changes implicate other ones. Also adjusting every backend per commit would be really painful. I think this PR should be treated as adding support for the Host Commands from scratch, because of amount of changes. The previous system wasn't usable in my opinion.

keith-zephyr commented:
The file move of `subsys/mgmt/ec_host_cmd/ec_host_cmd_periph/ec_host_cmd_periph_shi.h` needs to update the #ifdef guard for the new path.

keith-zephyr commented on include/zephyr/mgmt/ec_host_cmd/backend.h:
```diff
@@ -21,98 +21,110 @@
- * receive data from the host.
+ * This routine initializes a host command backend. It includes initialization
+ * a device used to communication and setting up buffers.
+ * This function is called by the ec_host_cmd_init function. However, if
+ * DT_CHOSEN(zephyr_host_cmd_backend) is defined, the ec_host_cmd_init function
+ * should be called automatically by a backend using SYS_INIT with
+ * CONFIG_EC_HOST_CMD_INIT_PRIORITY priority.
```
No need to rehash the ec_host_cmd_init() requirements again.
```suggestion
 * This function is called by the ec_host_cmd_init function.
```

keith-zephyr commented on subsys/mgmt/ec_host_cmd/Kconfig:
```diff
@@ -9,20 +9,39 @@ config EC_HOST_CMD
 	default 256
+	help
+	  Buffer size in bytes for TX buffer defined by the host command handler.
+	  Some backend layers can define their own buffer, so the size can be zero not to
+	  duplicate buffers. If multiple backends are used, the size has to be set by user
```
```suggestion
	  Some backend layers can define their own buffer, so the size can be zero to avoid
	  duplicate buffers. If multiple backends are used, the size has to be set by user
```

keith-zephyr commented on subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_espi.c:
```diff
@@ -4,103 +4,133 @@
-	rx_ctx->handler_owns = &data->handler_owns;
+	/* eSPI backend uses the shared memory as tx buffer */
+	if (tx->buf) {
+		LOG_WRN("Providing unused buffer");
```
Move this check (and other backends) into ec_host_cmd_init() so you have a consistent warning message regardless of the backend selected.

keith-zephyr commented on subsys/mgmt/ec_host_cmd/ec_host_cmd_handler.c:
```diff
@@ -133,79 +180,56 @@ static void handle_host_cmds_entry(void *arg1, void *arg2, void *arg3)
 
-		ec_host_cmd_backend_send(ec_host_cmd_dev, &tx);
+	ret = backend->api->init(backend, backend_config, &hc->rx_ctx, &hc->tx);
+	if (ret != 0) {
```
Verify that the buffers are not NULL. Also check if the buffers changed and log a warning so the user can eliminate the duplicate buffer.

keith-zephyr commented on subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c:
```diff
@@ -184,18 +193,18 @@ static int shi_ite_backend_send(const struct device *dev,
+	 */
 	for (int i = 0; i < EC_SHI_PAST_END_LENGTH; i++) {
-		data->out_msg[sizeof(out_preamble) + buf->len + i] = EC_SHI_PAST_END;
+		data->out_msg[sizeof(out_preamble) + data->tx->len + i] = EC_SHI_PAST_END;
```
Can you use memset() here instead?

keith-zephyr commented on subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_espi.c:
```diff
@@ -4,103 +4,133 @@
+	if (rx_header->prtcl_ver != 3 ||
+	    rx_valid_data_size > CONFIG_EC_HOST_CMD_HANDLER_RX_BUFFER ||
+	    rx_valid_data_size > shared_size) {
+		memcpy(hc_espi->rx_ctx->buf, (void *)rx_header, RX_HEADER_SIZE);
```
Don't you need a `k_sem_take()` before writing into the RX buffer?  Can the host send another command while the handler is still processing a command?

semihalf-niedzwiecki-dawid replied:
Done.

semihalf-niedzwiecki-dawid commented:
> The file move of `subsys/mgmt/ec_host_cmd/ec_host_cmd_periph/ec_host_cmd_periph_shi.h` needs to update the #ifdef guard for the new path.

Done

semihalf-niedzwiecki-dawid commented:
Friendly ping :-)

keith-zephyr commented on doc/services/device_mgmt/ec_host_cmd.rst:
```diff
@@ -0,0 +1,64 @@
+
+Overview
+********
+EC Host Command is a subsystem used to communicate with a host via the host command protocol.
+Only version 3 of the protocol is supported.
```
"The host command protocol defines the interface for a host, or application processor, to communicate with a target embedded controller (EC).  The EC Host command subsystem implements the target side of the protocol, generating responses to commands sent by the host.   The host command protocol interface supports multiple versions, but this subsystem implementation only support protocol version 3."

semihalf-niedzwiecki-dawid replied:
Done.

keith-zephyr commented on subsys/mgmt/ec_host_cmd/Kconfig:
```diff
@@ -9,20 +9,39 @@ config EC_HOST_CMD
-	  computers. Enabling this option requires specifying a chosen
-	  zephyr,ec-host-interface device as the ec host command backend that
-	  receive incoming host command requests to process.
+	  computers. Only version 3 is supported.
```
Drop "Only version 3 is supported" because this is noted in the documentation.

semihalf-niedzwiecki-dawid replied:
Done.

semihalf-niedzwiecki-dawid commented:
Sorry, I've added many comments that haven't been visible for you.

semihalf-niedzwiecki-dawid commented on include/zephyr/mgmt/ec_host_cmd/ec_host_cmd.h:
```diff
@@ -213,6 +225,23 @@ enum ec_host_cmd_status {
+ * This routine initializes the host command subsystem. It includes initialization
+ * of a backend and the handler.
+ * When the application configures the zephyr,host-cmd-backend chosen node, the chosen backend
+ * automatically calls this routing at CONFIG_EC_HOST_CMD_INIT_PRIORITY.
```
Done.

semihalf-niedzwiecki-dawid commented on include/zephyr/mgmt/ec_host_cmd/backend.h:
```diff
@@ -21,98 +21,107 @@
-	 *  read from this buffer (for security reasons).
+	/**
+	 * Buffer to hold received data. The buffer is provided by the handler if
+	 * EC_HOST_CMD_HANDLER_RX_BUFFER > 0. Otherwise, the backend should provide
```
Done.

semihalf-niedzwiecki-dawid commented on include/zephyr/mgmt/ec_host_cmd/backend.h:
```diff
@@ -21,98 +21,107 @@
+	/** API provided by the backed. */
+	const struct ec_host_cmd_backend_api *api;
+	/** Context for the backed. */
+	void *ctx;
```
The same answer for all `struct device *` vs `void *` config. I'm implementing UART and SPI backends for the FPMCU. At the moment, the SPI support needs more than struct device * as config, so would prefer to keep `void *`. If it turns out that `struct device *` is only one required config parameter for all backends, I will changes that, ok?

semihalf-niedzwiecki-dawid replied:
Anyway, the `void * ctx` pointer is something different, than the config. It is a pointer to a context for backend. It points to struct ec_host_cmdbackendctx

semihalf-niedzwiecki-dawid commented on include/zephyr/mgmt/ec_host_cmd/ec_host_cmd.h:
```diff
@@ -213,6 +225,21 @@ enum ec_host_cmd_status {
+ *
+ * @retval 0 if successful
+ */
+int ec_host_cmd_init(struct ec_host_cmd_backend *backend, const void *backend_config);
```
The same answer for all struct device * vs void * config. I'm implementing UART and SPI backends for the FPMCU. At the moment, the SPI support needs more than struct device * as config, so would prefer to keep void *. If it turns out that struct device * is only one required config parameter for all backends, I will changes that, ok?

semihalf-niedzwiecki-dawid commented on subsys/mgmt/ec_host_cmd/Kconfig:
```diff
@@ -9,20 +9,39 @@ config EC_HOST_CMD
+	help
+	  Buffer size in bytes for TX buffer defined by the host command handler.
+	  Some backend layers can define their own buffer, so the size can be zero to
+	  duplilate buffers. If multiple backends are used, the size has to be set by user
```
Done.

semihalf-niedzwiecki-dawid commented on subsys/mgmt/ec_host_cmd/Kconfig:
```diff
@@ -9,20 +9,39 @@ config EC_HOST_CMD
+	help
+	  Buffer size in bytes for TX buffer defined by the host command handler.
+	  Some backend layers can define their own buffer, so the size can be zero to
+	  duplicate buffers. If multiple backends are used, the size has to be set by user
```
Done.

semihalf-niedzwiecki-dawid commented on subsys/mgmt/ec_host_cmd/ec_host_cmd_handler.c:
```diff
@@ -133,79 +185,71 @@ static void handle_host_cmds_entry(void *arg1, void *arg2, void *arg3)
 
-		struct ec_host_cmd_response_header *const tx_header =
-			(void *)tx_buffer;
+	hc->backend = backend;
```
`ctx` is a pointer to a backend context structure. It is assigned by the backend in the init function. The `backend_config` is passed as an argument to the init function and can be stored by the backend or just used to init peripheral.

keith-zephyr replied:
Ack - I see ctx is initialized by each backend.

semihalf-niedzwiecki-dawid commented on include/zephyr/mgmt/ec_host_cmd/ec_host_cmd.h:
```diff
@@ -213,6 +225,23 @@ enum ec_host_cmd_status {
+ * zephyr,host-cmd-backend undefined and must explicitly call this routine.
+ *
+ * @param[in] backend        Pointer to the backend structure to initialize.
+ * @param[in] backend_config Pointer to backend dependent config data
```
I don't think it is a good idea. The application code needs a way to pass some kind of config to initialize backend. The backend context structures are only available inside the backend files. Anyway, the `ctx` pointer isn’t assigned before the `ec_host_cmd_backend_api->init` call. IMO, the separated config argument is the best option here. The shell initialization also uses this pattern.

keith-zephyr replied:
The `backend_config` data varies based on the backend, so the application has to have insider knowledge about what to pass here.

For the the SHI backends, the `backend_config` is fixed - it must always be the `struct device` for the SHI driver.  So these backends can statically assign `ctx->dev = DEVICE_DT_INST_GET(0)`.

For the eSPI backend, how about adding a `struct device *` parameter to `ec_host_cmd_backend_get_espi()` to specify the driver to use.  This routine would then initialize `ctx->espi_dev`. 

This allows you to document exactly what `ec_host_cmd_backend_get_espi()` requires,

semihalf-niedzwiecki-dawid replied:
Done. Documenting config per backend sounds good.

keith-zephyr commented on subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_espi.c:
```diff
@@ -4,103 +4,161 @@
+	    rx_valid_data_size > shared_size) {
+		memcpy(hc_espi->rx_ctx->buf, (void *)rx_header, RX_HEADER_SIZE);
+		hc_espi->rx_ctx->len = RX_HEADER_SIZE;
+		hc_espi->state = ESPI_STATE_RX_BAD;
```
You're not using this state anywhere else, you should just set the state to `ESPI_STATE_PROCESSING` since you're passing the request to the general handler anyway.

semihalf-niedzwiecki-dawid replied:
Done.

keith-zephyr commented on subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_espi.c:
```diff
@@ -4,103 +4,159 @@
+	ESPI_STATE_DISABLED,
+	/* Ready to receive next request */
+	ESPI_STATE_READY_TO_RECV,
+	/* Received bad data */
+	ESPI_STATE_RX_BAD,
```
This can be removed.

semihalf-niedzwiecki-dawid replied:
Done.

keith-zephyr reviewed (commented):
One minor change requested.
